### PR TITLE
fix(WebSocketShard): Zombie connection fix

### LINKED ISSUE #8486: Bot randomly exiting process or going offline/unresponsive
### Which package is this bug report for?

discord.js

### Issue description

After around 5-6 hours of running, my discord bot decides to randomly exit with an exit code of 0. There is nothing in my code that should cause that, and adding a debug event listener shows the bot failing around a websocket reacquisition near the 6 hour mark.

I have tried running my bot with typescript, I've tried compiling the code to javascript and running that, I've tried packaging it as a docker container and running that locally, I've tried packing it as a docker container and running that on my production server. Every environment and method results in the same issue.

image

### Bot code
https://github.com/NeonWizard/sockbot-discord

### Logs
```
[WS => Shard 0] Heartbeat acknowledged, latency of 86ms.
[WS => Shard 0] [HeartbeatTimer] Sending a heartbeat.
[WS => Shard 0] Heartbeat acknowledged, latency of 89ms.
[WS => Shard 0] [HeartbeatTimer] Sending a heartbeat.
[WS => Shard 0] Heartbeat acknowledged, latency of 85ms.
[WS => Shard 0] [HeartbeatTimer] Sending a heartbeat.
[WS => Shard 0] Heartbeat acknowledged, latency of 83ms.
[WS => Shard 0] [HeartbeatTimer] Sending a heartbeat.
[WS => Shard 0] Heartbeat acknowledged, latency of 84ms.
[WS => Shard 0] [RECONNECT] Discord asked us to reconnect
[WS => Shard 0] [DESTROY]
    Close Code    : 4000
    Reset         : false
    Emit DESTROYED: true
[WS => Shard 0] Clearing the heartbeat interval.
[WS => Shard 0] [WebSocket] Destroy: Attempting to close the WebSocket. | WS State: OPEN      
[WS => Shard 0] [WebSocket] Close: Tried closing. | WS State: CLOSING
[WS => Shard 0] [WebSocket] Adding a WebSocket close timeout to ensure a correct WS reconnect.
        Timeout: 5000ms
[WS => Shard 0] [WebSocket] Clearing the close timeout.
[WS => Shard 0] [WebSocket] Close Emitted: false
[WS => Shard 0] [WebSocket] did not close properly, assuming a zombie connection.
Emitting close and reconnecting again.
[WS => Shard 0] [CLOSE]
    Event Code: 1011
    Clean     : false
    Reason    : INTERNAL_ERROR
[WS => Shard 0] Session id is present, attempting an immediate reconnect...
[WS => Shard 0] [CONNECT]
    Gateway    : wss://gateway.discord.gg/
    Version    : 10
    Encoding   : json
    Compression: none
[WS => Shard 0] Setting a HELLO timeout for 20s.
[WS => Shard 0] [CONNECTED] Took 146ms
[WS => Shard 0] Clearing the HELLO timeout.
[WS => Shard 0] Setting a heartbeat interval for 41250ms.
[WS => Shard 0] [RESUME] Session REDACTED, sequence 51
[WS => Shard 0] [RESUMED] Session REDACTED | Replayed 1 events.       
[WS => Shard 0] [ResumeHeartbeat] Sending a heartbeat.
[WS => Shard 0] Heartbeat acknowledged, latency of 88ms.
[WS => Shard 0] Clearing the heartbeat interval.
[WS => Shard 0] [CLOSE]
    Event Code: 4000
    Clean     : true
    Reason    :
[WS => Shard 0] Session id is present, attempting an immediate reconnect...
[WS => Shard 0] An open connection was found, attempting an immediate identify.
[WS => Shard 0] [RESUME] Session REDACTED, sequence 52
root@ruby:/home/spooky#
```

### Code sample

```typescript
https://github.com/NeonWizard/sockbot-discord
```

### Package version

discord.js@14.1.2

### Node.js version

v16.16.0, typescript@4.7.4

### Operating system

Windows (WSL), Linux (Ubuntu), Docker (node:lts-alpine)

### Priority this issue should have

High (immediate attention needed)

### Which partials do you have configured?

No Partials

### Which gateway intents are you subscribing to?

Guilds, GuildMessages, GuildMessageReactions, MessageContent

### I have tested this issue on a development release

_No response_

### LINKED ISSUE #8984: bot doesn't reconnect after reidentifying
### Which package is this bug report for?

discord.js

### Issue description

here's the debug log: 
[2022-12-27-debug.log](https://github.com/discordjs/discord.js/files/10309867/2022-12-27-debug.log)
at the end of it you can see that it tries to identify but it never actually happens

### Code sample

I'm currently on phone so here's the [file](https://github.com/dulcisvtc/webapi/blob/master/src/index.ts)

### Package version

14.7.1

### Node.js version

18.12.0

### Operating system

Linux 42dde76b-d8e7-4868-921f-8c263d171b72 5.15.0-1013-oracle #17~20.04.1-Ubuntu SMP Mon Jul 4 05:29:46 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux

### Priority this issue should have

Medium (should be fixed soon)

### Which partials do you have configured?

No Partials

### Which gateway intents are you subscribing to?

Guilds, GuildMembers, GuildMessages, MessageContent

### I have tested this issue on a development release

_No response_

### PR DESCRIPTION
**Pull request by legendhimself**
Fixes #8486, fixes #8984 and supersedes #8981 

- Fix backport #7626 missing changes
- Reverted the pull request #8956
- Removed unref of wsCloseTimeout
- ~~We are resuming the connection for zombie instead of starting a new~~ To not mess it up we didn't push this. Will create a separate pr for this.

**Please describe the changes this PR makes and why it should be merged:**

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating

### CODE REVIEW

**Comment by legendhimself** [comment]
<details>
<summary>
Previous Report
</summary>

image
</details>

Before the uptime was kind of bad thanks to the `unref` that we had on the `WsCloseTimeout`,

<details>
<summary>
Current Report
</summary>

image
</details>

The logs/tests proved reduced reconnects and better uptime at least on my 60+ Shards bot.
@DraftProducts Also has tested the changes on his bot with 350+ shards which had frequent zombie connections and most of the time ran out of logins per day.

Also thanks @ShoXcy for providing me with shawarma during the testings

**Comment by kyranet** [comment]
Blocking for 2 days at @legendhimself's request so they can test the PR for more time. Will unblock afterwards.

**Comment by DraftProducts** [comment]
For now, I have excellent results.
Previously I was loosing multiple starts bucket tokens each 5 mins that brought me to multiple bot's token resets per day (3 to 5 times with 2000 bucket logins), had zombies connections and offline status bot on others. 
I didn't notice any re-login or zombie connections since 2am when I merged our fix in production.

**Comment by DraftProducts** [comment]
I hear that you were waiting for more information, here are my session_start_limit remaining logins of the GET [`/bot/gateway`](https://discord.com/developers/docs/topics/gateway#get-gateway-bot) request (without any token reset today).
```json
{
  "url": "wss://gateway.discord.gg",
  "shards": 368,
  "session_start_limit": {
  	"total": 2000,
  	"remaining": 1995,
  	"reset_after": 79901986,
  	"max_concurrency": 16
  }
}
```

That shows that we just used 5 relogins today without any zombie connection, so for me the problem is solved, **and this PR should be merged as soon as possible for others devs** 🎉

**Review by vladfrangu — APPROVED** [review]

**Comment by legendhimself** [comment]
Day 2,
<details>
<summary>
Logins today
</summary>

image
</details>

<details>
<summary>
Stats Report
</summary>

image
</details>
Uptime ^, No restarts so far. Usually, some of the shards by this time used to silently restart due to the unref on the closetimeout but now, as you saw in the previous login, everything is going fine with this fix.

**Review by kyranet — APPROVED** [review]

File: packages/discord.js/src/client/websocket/WebSocketShard.js [hunk 1]
```diff
@@ -837,25 +840,19 @@ class WebSocketShard extends EventEmitter {
...
     } else if (emit) {
       // We requested a destroy, but we had no connection. Emit destroyed
       this._emitDestroyed();
     }
 
+    this.debug(
+      `[WebSocket] Adding a WebSocket close timeout to ensure a correct WS reconnect.
+        Timeout: ${this.manager.client.options.closeTimeout}ms`,
+    );
+    this.setWsCloseTimeout(this.manager.client.options.closeTimeout);
```
Comment (Qjuh) [inline packages/discord.js/src/client/websocket/WebSocketShard.js:854]: Moving this out of the `if (this.connection.readyState === WebSocket.OPEN) {}` block reintroduces the issue where the timeout would wait for closes on websockets that already were closed before and try to reconnect after we already successfully reconnected.
Either this should go back inside the if or at least the fix of my PR you supersede here to clear the timeout in the connect method before connecting (if one still exists) should be added in this PR too.

**Review by Qjuh — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/client/websocket/WebSocketShard.js (hunk 1 above)
Comment (legendhimself) [reply packages/discord.js/src/client/websocket/WebSocketShard.js:854]: > the timeout would wait for closes on WebSockets that already were closed before
I have tested this case already, `this.closeEmitted` will clear the timeout if the close was ever emitted before.

Comment (legendhimself) [reply packages/discord.js/src/client/websocket/WebSocketShard.js:854]: <details>
<summary>
Logs
</summary>

```js
[ DJS ]: | Sun, 01 Jan 2023 12:57:20 | debug | [WS => Shard 25] | [RECONNECT] Discord asked us to reconnect
[ DJS ]: | Sun, 01 Jan 2023 12:57:20 | debug | [WS => Shard 25] | [DESTROY]
    Close Code    : 4000
    Reset         : false
    Emit DESTROYED: true
[ DJS ]: | Sun, 01 Jan 2023 12:57:20 | debug | [WS => Shard 25] | Clearing the heartbeat interval.
[ DJS ]: | Sun, 01 Jan 2023 12:57:20 | debug | [WS => Shard 25] | [WebSocket] Destroy: Attempting to close the WebSocket. | WS State: OPEN
[ DJS ]: | Sun, 01 Jan 2023 12:57:20 | debug | [WS => Shard 25] | [WebSocket] Close: Tried closing. | WS State: CLOSING
[ DJS ]: | Sun, 01 Jan 2023 12:57:20 | debug | [WS => Shard 25] | [WebSocket] Adding a WebSocket close timeout to ensure a correct WS reconnect.
        Timeout: 2500ms
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | [WebSocket] Clearing the close timeout.
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | [CLOSE]
    Event Code: 4000
    Clean     : true
    Reason    : 
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | Session id is present, attempting an immediate reconnect...
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | [CONNECT]
    Gateway    : wss://gateway.discord.gg/
    Version    : 9
    Encoding   : json
    Compression: none
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | Setting a HELLO timeout for 20s.
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | [CONNECTED] Took 198ms
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | Clearing the HELLO timeout.
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | Setting a heartbeat interval for 41250ms.
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | [RESUME] Session ~~, sequence 1384607
[ DJS ]: | Sun, 01 Jan 2023 12:57:21 | debug | [WS => Shard 25] | [RESUMED] Session ~~ | Replayed 1 events.
```
</details>

This happened 804 times in the past 3 days 13 hrs.

Also, the issue you were talking about is this #8984 from @itzJOHv and this should be fixed with this pr.

Comment (legendhimself) [reply packages/discord.js/src/client/websocket/WebSocketShard.js:854]: For **testing**, you can also try to create a 4009 (zombie connection) yourself by not acknowledging the heartbeat from our side. Or the other way around, don't send the heartbeat which will cause a zombie connection again.
```js
  ackHeartbeat() {
    this.lastHeartbeatAcked = false;
```
<details>
<summary>
Logs
</summary>

```
//  node .\packages\discord.js\test\shard.js
kek dying
[WS => Shard 0] [HeartbeatTimer] Didn't receive a heartbeat ack last time, assuming zombie connection. Destroying and reconnecting.
    Status          : 0
    Sequence        : 2
    Connection State: OPEN
[WS => Shard 0] [DESTROY]
    Close Code    : 4009
    Reset         : true
    Emit DESTROYED: true
[WS => Shard 0] Clearing the heartbeat interval.
[WS => Shard 0] [WebSocket] Destroy: Attempting to close the WebSocket. | WS State: OPEN
[WS => Shard 0] [WebSocket] Close: Tried closing. | WS State: CLOSING
[WS => Shard 0] [WebSocket] Adding a WebSocket close timeout to ensure a correct WS reconnect.
        Timeout: 5000ms
[WS => Shard 0] [WebSocket] Clearing the close timeout.
[WS => Shard 0] [CLOSE]
    Event Code: 4009
    Clean     : true
    Reason    :
[WS => Shard 0] [CONNECT]
    Gateway    : wss://gateway.discord.gg/
    Version    : 10
    Encoding   : json
    Compression: zlib-stream
[WS => Shard 0] Setting a HELLO timeout for 20s.
[WS => Shard 0] [CONNECTED] Took 353ms
[WS => Shard 0] Clearing the HELLO timeout.
[WS => Shard 0] Setting a heartbeat interval for 41250ms.
[WS => Shard 0] [IDENTIFY] Shard 0/1 with intents: 513
[WS => Shard 0] [READY] Session a90b802c577369648e9a96d97c54994f | Resume url wss://gateway-us-east1-d.discord.gg.
[WS => Shard 0] [ReadyHeartbeat] Sending a heartbeat.
[WS => Shard 0] Shard received all its guilds. Marking as fully ready.
[WS => Shard 0] Heartbeat acknowledged, latency of 303ms.
```
</details>
And doing a zombie connection from our side did a successful reconnect.

Before I had set `this.closeEmitted` as false just before initiating a WsCloseTimeout which caused the issue that you are trying to fix in #8981
But now `this.closeEmitted` is set only twice, once in the onClose `this.closeEmitted = true` and once in the WsCloseTimeout `this.closeEmitted = false`

If you are still unsure whether this pr is going to cause issues please provide me the tests/logs in which it does

Comment (legendhimself) [reply packages/discord.js/src/client/websocket/WebSocketShard.js:854]: One can create these kind of conditions by editing the Ws code accordingly.

Edge cases that I tested with this pr:
- not ack from our side
- not being ack from the discord side (not sending the ping to discord)
- close never being invoked
- all 3 above combined

Comment (legendhimself) [reply packages/discord.js/src/client/websocket/WebSocketShard.js:854]: > websocket connection being already closed
Won't the `closeEmitted` be true if the websocket already was closed? https://github.com/discordjs/discord.js/blob/a7fd2d8194677d6e90cda96194b29ace9f8238af/packages/discord.js/src/client/websocket/WebSocketShard.js#L595

This will clear the timeout and return early.

If you can provide me a minimal repro using this pr so I can look into it.

Comment (legendhimself) [reply packages/discord.js/src/client/websocket/WebSocketShard.js:854]: Also, the logs that you gave have error code 1011 which is not from the current code(pr). It will be helpful if you can get the exact same results with this current pr.

**Review by iCrawl — APPROVED** [review]
Looking good

File: packages/discord.js/src/client/websocket/WebSocketShard.js (hunk 1 above)
Comment (iCrawl) [reply packages/discord.js/src/client/websocket/WebSocketShard.js:854]: We should be fine sorting this out on a follow-up if anything comes up.
